dmd/expression.d

```diff
@@ -3732,6 +3787,13 @@ extern (C++) final class VarExp : SymbolExp
             error("cannot modify operator `$`");
             return ErrorExp.get();
         }
+        if (auto fd = var.isFuncDeclaration()) {
+            auto fty = fd.type.toTypeFunction();
+            if (fty.isCtonly()) {
+                error("cannot take address of CTFE-only function `%s`", var.toChars());
```

Comment:
This is more complicated than it seems, because under CTFE context, this should be allowed, you should guard the code with `sc.flags & SCOPE.ctfe`, but I think its not enough, because it would only guard for CTFE instances.

Maybe do a test with taking addresses inside CTFE and on functions with `@__ctfe` that are also only `@__ctfe`. So this check should also contemplate that case.
I would go even further, why wouldn't we be able to take the address of a non-`@__ctfe` function, if that address is only used in the local scope?

In general, I believe this check should be elsewhere, near the fe/be glue code, because, taking address should be forbidden only when its stored in globals, no?

Reply:
I think it's totally fine to stay on the conservative side here. What **is** the address of a CT-only function anyway? Yes, we can support some cases with first taking an address inside CTFE and then calling it using that address also inside CTFE. But that requires more complicated analysis: what if an address is casted to an integer, put into a structure and returned? It's hard to detect if it's going to leak to run-time.